core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala

```diff
@@ -0,0 +1,552 @@
+      } else {
+        return KillResponse(submissionId, false, "Cannot find driver")
+      }
+    }
+  }
+
+  def recoverState {
+    stateLock.synchronized {
+      queue = new DriverQueue(engineFactory.createEngine("driverQueue"), queuedCapacity)
+
+      launchedDrivers = new LaunchedDrivers(engineFactory.createEngine("launchedDrivers"))
+
+      // There is potential timing issue where a queued driver might have been launched
+      // but the scheduler shuts down before the queued driver was able to be removed
+      // from the queue. We try to mitigate this issue by walking through all queued drivers
+      // and remove if they're already launched.
+      queue.drivers.foreach {
+        d => if (launchedDrivers.contains(d.submissionId)) {
+          queue.remove(d.submissionId)
+        }
+      }
+
+      superviseRetryList = new SuperviseRetryList(engineFactory.createEngine("retryList"))
+
+      // TODO: Consider storing finished drivers so we can show them on the UI after
+      // failover. For now we clear the history on each recovery.
+      finishedDrivers.clear()
+    }
+  }
+
+  def start() {
+    // TODO: Implement leader election to make sure only one framework running in the cluster.
+    val fwId = state.fetch[String]("frameworkId")
+
+    val builder = FrameworkInfo.newBuilder()
+      .setUser(Utils.getCurrentUserName())
+      .setName(appName)
+      .setWebuiUrl(frameworkUrl)
+      .setCheckpoint(true)
+      .setFailoverTimeout(Integer.MAX_VALUE) // Setting to max for tasks keep running until recovery
+
+    fwId.foreach { id =>
+      builder.setId(FrameworkID.newBuilder().setValue(id).build())
+      appId = id
+    }
+
+    // Recover scheduler state that is persisted.
+    // We still need to do task reconciliation to be up to date of the latest task states
+    // as it might have changed while the scheduler is failing over.
+    recoverState
+    startScheduler("MesosClusterScheduler", master, MesosClusterScheduler.this, builder.build())
+  }
+
+  def stop() {
+    driver.stop(true)
+  }
+
+  override def registered(
+      driver: SchedulerDriver,
+      frameworkId: FrameworkID,
+      masterInfo: MasterInfo): Unit = {
+    logInfo("Registered as framework ID " + frameworkId.getValue)
+    if (frameworkId.getValue != appId) {
+      appId = frameworkId.getValue
+      state.persist("frameworkId", appId)
+    }
+    markRegistered()
+
+    stateLock.synchronized {
+      if (!launchedDrivers.pendingRecover.isEmpty) {
+        // Start task reconciliation if we need to recover.
+        val statuses = launchedDrivers.pendingRecover.collect {
+          case (taskId, slaveId) =>
+            launchedDrivers.get(taskId).taskState.getOrElse(
+              TaskStatus.newBuilder()
+                .setTaskId(TaskID.newBuilder().setValue(taskId).build())
+                .setSlaveId(slaveId)
+                .setState(TaskState.TASK_STAGING)
+                .build)
```

Comment:
not a getter, `build()` (same with `queue.poll()` and other things in this file)